SpatialGDK/Source/SpatialGDK/Public/Interop/Connection/SpatialEventTracer.h

```diff
@@ -0,0 +1,108 @@
+// Note(EventTracer): SpatialEventTracer is supposed to never be null in SpatialWorkerConnection, SpatialSender, SpatialReceiver. Make sure
+// there are necessary nullptr checks if that changes.
+
+class SpatialEventTracer
+{
+public:
+	SpatialEventTracer(const FString& WorkerId);
+	~SpatialEventTracer();
+
+	const worker::c::Trace_EventTracer* GetConstWorkerEventTracer() const { return EventTracer; };
+	worker::c::Trace_EventTracer* GetWorkerEventTracer() const { return EventTracer; }
+
+	// TODO(EventTracer): add the option to SpatialScopedActiveSpanId for Sent TraceEvents.
+	// Consider making sure it's not accepting rvalue (since SpatialSpanIdActivatNetDriveror must live long enough for the worker sent op to
+	// be registered with this SpanId) e.g. void TraceEvent(... SpatialScopedActiveSpanId&& SpanIdActivator) = delete;
+	// TODO(EventTracer): Communicate to others, that SpatialScopedActiveSpanId must be creating prior to calling worker send functions
+
+	TOptional<Trace_SpanId> TraceEvent(FSpatialTraceEvent SpatialTraceEvent);
+	TOptional<Trace_SpanId> TraceEvent(FSpatialTraceEvent SpatialTraceEvent, const worker::c::Trace_SpanId Causes);
+	TOptional<Trace_SpanId> TraceEvent(FSpatialTraceEvent SpatialTraceEvent, const TArray<worker::c::Trace_SpanId>& Causes);
+	TOptional<Trace_SpanId> TraceEvent(FSpatialTraceEvent SpatialTraceEvent, const worker::c::Trace_SpanId* Causes, int32 NumCauses);
+
+	bool IsEnabled() const;
+
+	void ComponentAdd(const Worker_Op& Op);
+	void ComponentRemove(const Worker_Op& Op);
+	void ComponentUpdate(const Worker_Op& Op);
+
+	bool GetSpanId(const EntityComponentId& Id, const uint32 FieldId, worker::c::Trace_SpanId& CauseSpanId, bool bRemove = true);
+	bool GetMostRecentSpanId(const EntityComponentId& Id, worker::c::Trace_SpanId& CauseSpanId, bool bRemove = true);
+
+	bool DropSpanId(const EntityComponentId& Id, const uint32 FieldId);
+	bool DropSpanIds(const EntityComponentId& Id);
+
+private:
+	bool bEnabled{ false };
+	worker::c::Trace_EventTracer* EventTracer{ nullptr };
+
+	uint64 BytesWrittenToStream{ 0 };
+	uint64 MaxFileSize{ 0 };
+
+	struct StreamDeleter
+	{
+		void operator()(worker::c::Io_Stream* StreamToDestroy) const;
+	};
+
+	TUniquePtr<worker::c::Io_Stream, StreamDeleter> Stream;
+
+	SpatialSpanIdCache SpanIdStore;
+
+	void Enable(const FString& FileName);
+	void Disable();
+
+	static void TraceCallback(void* UserData, const Trace_Item* Item);
+};
+
+struct SpatialScopedActiveSpanId
+{
+	SpatialScopedActiveSpanId(SpatialEventTracer* InEventTracer, const TOptional<Trace_SpanId>& InCurrentSpanId);
+	~SpatialScopedActiveSpanId();
+
+	SpatialScopedActiveSpanId(const SpatialScopedActiveSpanId&) = delete;
+	SpatialScopedActiveSpanId(SpatialScopedActiveSpanId&&) = delete;
+	SpatialScopedActiveSpanId& operator=(const SpatialScopedActiveSpanId&) = delete;
+	SpatialScopedActiveSpanId& operator=(SpatialScopedActiveSpanId&&) = delete;
+
+private:
+	TOptional<Trace_SpanId> CurrentSpanId;
```

Comment:
I don't think you need to keep the current value around. I'd say that you have to provide a valid span ID or if you want this to optionally do nothing then make this a const ref as it's a non movable type anyway.

Reply:
What happens if you call clear active span while there is none?